src/main/java/com/radixdlt/crypto/encryption/ECIES.java

```diff
@@ -0,0 +1,167 @@
+import com.radixdlt.crypto.ECPublicKey;
+import com.radixdlt.crypto.Hash;
+import com.radixdlt.crypto.MacMismatchException;
+import org.bouncycastle.math.ec.ECPoint;
+
+import javax.crypto.Mac;
+import javax.crypto.spec.SecretKeySpec;
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.DataInputStream;
+import java.io.DataOutputStream;
+import java.io.IOException;
+import java.security.GeneralSecurityException;
+import java.security.SecureRandom;
+import java.util.Arrays;
+import java.util.Random;
+
+/**
+ * Encrypt and Decrypt data using ECIES
+ * (Elliptic Curve <a href="https://en.wikipedia.org/wiki/Integrated_Encryption_Scheme">Integrated Encryption Scheme</a>, subset of DHIES):
+ */
+public final class ECIES {
+	private ECIES() {
+		// cannot construct
+	}
+
+	public static <M extends ECMultiplicationScalar> byte[] decrypt(byte[] data, M multiplicationScalar) throws ECIESException {
+		try {
+			DataInputStream inputStream = new DataInputStream(new ByteArrayInputStream(data));
+
+			// 1. Read the IV
+			byte[] iv = new byte[16];
+			inputStream.readFully(iv);
+
+			// 2. Read the ephemeral public key
+			int publicKeySize = inputStream.readUnsignedByte();
+			byte[] publicKeyRaw = new byte[publicKeySize];
+			inputStream.readFully(publicKeyRaw);
+			ECPublicKey ephemeralPublicKey = new ECPublicKey(publicKeyRaw);
+
+			// 3. Do an EC point multiply with this.getPrivateKey() and ephemeral public key. This gives you a point M.
+			ECPoint m = multiplicationScalar.multiply(ephemeralPublicKey.getPublicPoint());
+
+			// 4. Use the X component of point M and calculate the Hash H.
+			byte[] h = hash(m.getXCoord().getEncoded());
+
+			// 5. The first 32 bytes of H are called key_e and the last 32 bytes are called key_m.
+			byte[] keyE = Arrays.copyOfRange(h, 0, 32);
+			byte[] keyM = Arrays.copyOfRange(h, 32, 64);
+
+			// 6. Read encrypted data
+			byte[] encrypted = new byte[inputStream.readInt()];
+			inputStream.readFully(encrypted);
+
+			// 6. Read MAC
+			byte[] mac = new byte[32];
+			inputStream.readFully(mac);
+
+			// 7. Compare MAC with MAC'. If not equal, decryption will fail.
+			byte[] pkMac = calculateMAC(keyM, iv, ephemeralPublicKey, encrypted);
+			if (!Arrays.equals(mac, pkMac)) {
+				throw new MacMismatchException(pkMac, mac);
+			}
+
+			// 8. Decrypt the cipher text with AES-256-CBC, using IV as initialization vector, key_e as decryption key
+			//    and the cipher text as payload. The output is the padded input text.
+			return Crypt.decrypt(iv, encrypted, keyE);
+		} catch (Exception e) {
+			throw new ECIESException("Failed to decrypt", e);
+		}
+	}
+
+	public static byte[] encrypt(byte[] data, ECPublicKey publicKey) throws ECIESException {
+		return encrypt(data, publicKey.getPublicPoint());
+	}
+
+	private static byte[] encrypt(byte[] data, ECPoint publicKeyPointOnCurve) throws ECIESException {
+		try {
+			Random rand = new SecureRandom();
```

Comment:
Do we really need to re-create an expensive SecureRandom instance every time this is called?

Reply:
good catch, fixed, also added throwing of `IllegalState` in the constructor.